text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,172 @@
+
+**Service** - an independent web process running in the cloud that provides RESTful endpoints
+
+**Client** - represents a physical end-user or software interface e.g. CLI, web UI, etc.
+ 
+Make a list of the definitions that may be useful for those reviewing. Include phrases and words that buildpack authors or other interested parties may not be familiar with.
+
+# Motivation
+[motivation]: #motivation
+
+As the CNB project makes progress towards supporting a centralized Buildpack Registry, I anticipate the need to quickly search for and identify specific buildpacks.  To support this, I propose the creation of an API search service that customers could use directly and/or indirectly via a web dashboard or CLI.
+
+- What use cases does it support?  
+  - Customers looking to quickly indentify usefulful buildpacks to build their applications.
+- What is the expected outcome?  
+  - Customers will be able to quickly find relevant buildpacks for their use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+An external API service that exposes endpoints for retrieving buildpack metadata.  Initially, I envision this service providing a few GET endpoints to support plain text query searches, and individual buildpack data retrieval. 
+
+### Supported Endpoints
+
+- **GET /buildpacks**
+  
+  Retrives all the buildpacks in the registry
+
+- **GET /buildpacks?query=text**
+
+  Retrievs all the buildpacks that statisfy the search query.  For example:
+  ```
+  $ curl https://glacial-plateau-95350.herokuapp.com/buildpacks?query=projectriff
+  ```
+  ```json
+  [
+    {
+      "ns":"projectriff",
+      "name":"command-function",
+      "version":"1.4.1",
+      "yanked":false,
+      "addr":"gcr.io/projectriff/command-function@sha256:99f9054abb73635a9b251b61d3627a8ff86508c767f9d691c426d45e8758596f"
+    },
+    {
+      "ns":"projectriff",
+      "name":"java-function",
+      "version":"1.4.1",
+      "yanked":false,"addr":"gcr.io/projectriff/java-function@sha256:5eabea8f7b2c09074ec196fe0c321006fb5ad8f282cc918520286d8a0007196f"},
+    {
+      "ns":"projectriff",
+      "name":"node-function",
+      "version":"1.4.1",
+      "yanked":false,
+      "addr":"gcr.io/projectriff/node-function@sha256:194298b826c15bb079c59aed99968d7678a6e1f7a882c9d7f61811e0990717ba"},
+    {
+      "ns":"projectriff",
+      "name":"streaming-http-adapter",
+      "version":"1.4.0",
+      "yanked":false,
+      "addr":"gcr.io/projectriff/streaming-http-adapter@sha256:b202d9ec203e882ee7e3c599d9e867617f909c8b6123e4ce942af47db6e58c45"
+    }
+  ]
+  ```
+
+- **GET /buildpacks/:id**
+
+  Retrieves metadata for a specific buildpack.  This will contain more detailed data.
+  ```
+  $ curl https://glacial-plateau-95350.herokuapp.com/buildpacks/1
+  ```
+  ```json
+  {
+      "description": "The Command Function Buildpack is a Cloud Native Buildpack V3 that provides riff Command Function Invoker to functions",
+      "license": "MIT",
```

Comment:
`description` and `license` might be different on different buildpack images (i.e. different versions of the same ID).

Reply:
Good to know.  So then I would expect a user would see a list of all the versions for a specific buildpack, and some of those version may have a different `description` and `license` field.   I assume this could be helpful as it would shed some light on the evolution of the buildpack.